test/extensions/stats_sinks/hystrix/hystrix_test.cc

```diff
@@ -0,0 +1,360 @@
+}
+
+TEST_F(HystrixSinkTest, BasicFlow) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+
+  // Arbitrary numbers for testing. Make sure error > timeout.
+  uint64_t success_step = 44;
+  uint64_t error_step = 33;
+  uint64_t timeout_step = 22;
+  uint64_t rejected_step = 11;
+
+  for (uint64_t i = 0; i < (window_size_ + 1); i++) {
+    buffer.drain(buffer.length());
+    ON_CALL(cluster1_.error_5xx_counter_, value()).WillByDefault(Return((i + 1) * error_step));
+    ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return((i + 1) * success_step));
+    cluster1_.cluster_info_->stats().upstream_rq_timeout_.add(timeout_step);
+    cluster1_.cluster_info_->stats().upstream_rq_pending_overflow_.add(rejected_step);
+    sink_->flush(source_);
+  }
+
+  std::string rolling_map = sink_->printRollingWindows();
+  EXPECT_NE(std::string::npos, rolling_map.find(cluster1_name_ + ".total"));
+
+  // Check stream format and data.
+  validate_results(TestUtility::bufferToString(buffer), success_step, error_step, timeout_step, 0,
+                   rejected_step, window_size_);
+
+  // Check the values are reset.
+  buffer.drain(buffer.length());
+  sink_->resetRollingWindow();
+  sink_->flush(source_);
+  validate_results(TestUtility::bufferToString(buffer), 0, 0, 0, 0, 0, window_size_);
+}
+
+//
+TEST_F(HystrixSinkTest, Disconnect) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+
+  // Arbitrary numbers for testing. Make sure error > timeout.
+  uint64_t success_step = 1;
+
+  for (uint64_t i = 0; i < (window_size_ + 1); i++) {
+    buffer.drain(buffer.length());
+    ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return((i + 1) * success_step));
+    sink_->flush(source_);
+  }
+
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "rollingCountSuccess"),
+            std::to_string(success_step * window_size_));
+  EXPECT_NE(buffer.length(), 0);
+
+  // Disconnect.
+  buffer.drain(buffer.length());
+  sink_->unregisterConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Reconnect.
+  buffer.drain(buffer.length());
+  sink_->registerConnection(&callbacks_);
+  ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return(success_step));
+  sink_->flush(source_);
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+}
+
+TEST_F(HystrixSinkTest, AddAndRemoveClusters) {
+  InSequence s;
+
+  // Arbitrary numbers for testing. Make sure error > timeout.
+  uint64_t success_step = 5;
+  uint64_t error_step = 4;
+  uint64_t timeout_step = 3;
+  uint64_t timeout_retry_step = 2;
+  uint64_t rejected_step = 1;
```

Comment:
If these are important relative to one another, you could make that clearer in the code in a few ways: e.g. specify them in numeric order, or maybe even assign them from a post-increment accumulator, e.g.
```
uint64_t accum = 1;
uint64_t rejected_step = accum++;
uint64_t timeout_retry_step = accum++;
```
etc

But I'm not sure if that's better than explicit numbers.

Make them const though.

Reply:
I'll leave them as explicit numbers. Changed some values to make it more random. These numbers should not be dependent on each other, they represent traffic. The only constraint is errors>timeouts, since Envoy counts timeouts together with errors, and Hystrix needs the values separated.

Reply:
In fact, I shouldn't be using "step", but instead randomly increasing values. But it would make the test much more complicated, and not sure it will improve testing.

Reply:
OK looking forward to the refactored version.  Hopefully much smaller, and with numbers that are more transparent in their meaning.